components/script/dom/formdata.rs

```diff
@@ -70,20 +75,60 @@ impl FormData {
 
 impl FormDataMethods<crate::DomTypeHolder> for FormData {
     // https://xhr.spec.whatwg.org/#dom-formdata
-    fn Constructor(
+    fn Constructor<'a>(
         global: &GlobalScope,
         proto: Option<HandleObject>,
         can_gc: CanGc,
-        form: Option<&HTMLFormElement>,
+        form: Option<&'a HTMLFormElement>,
+        submitter: Option<&'a HTMLElement>,
     ) -> Fallible<DomRoot<FormData>> {
+        // Helper to validate the submitter
+        fn validate_submitter<'b>(
+            submitter: &'b HTMLElement,
+            form: &'b HTMLFormElement,
+        ) -> Result<FormSubmitterElement<'b>, Error> {
+            let submit_button = submitter
+                .downcast::<HTMLButtonElement>()
+                .map(FormSubmitterElement::Button)
+                .or_else(|| {
+                    submitter
+                        .downcast::<HTMLInputElement>()
+                        .map(FormSubmitterElement::Input)
+                })
+                .ok_or(Error::Type(
+                    "submitter is not a form submitter element".to_string(),
+                ))?;
+
+            // Step 1.1.1. If submitter is not a submit button, then throw a TypeError.
```

Comment:
As a side note, which does not matter for now: I usually skip the step numbering, because the numbering changes all the time, so I only write the prose of the step, which tends to be more stable over time.

Reply:
@gterzian That's probably something we should discuss, since many other reviewers ask for step numbers right now.